### PR TITLE
add collector grammar

### PR DESCRIPTION
**Pull request by nofun97**
Changes proposed in this pull request:
- Add collector grammar
- Add collector statements
- REST endpoints allows special character
- Application names are now more flexible

Tested against all sysl files in this repository
Before: Passes: 176 Fails: 57
After.  : Passes: 180 Fails: 53

@anz-bank/sysl-developers

### CODE REVIEW

File: pkg/parser/endpoints.wbnf [hunk 1]
```diff
@@ -17,6 +17,13 @@ field_type -> collection_type | (type_spec optional="?"? attribs?) ;
 
 collector -> '.. * <- *' ":" (SHORTCUT | %!Indented(collector_stmts));
 
-collector_stmts ->  (
-                        call=(target=APPLICATION_NAME "<-" target_endpoint=NAME) | action=APPLICATION_NAME
-                    ) attribs;
+collector_stmts -> collector_http_stmt attribs
```
Comment (anzdaddy) [inline pkg/parser/endpoints.wbnf]: `collector_stmt`

Comment (anzdaddy) [inline pkg/parser/endpoints.wbnf]: Should be able to refactor `attribs` so it appears once.

File: pkg/parser/endpoints.wbnf [hunk 2]
```diff
@@ -17,6 +17,13 @@ field_type -> collection_type | (type_spec optional="?"? attribs?) ;
...
+                    | collector_pubsub_call attribs
+                    | collector_call_stmt attribs
+                    | action=APPLICATION_NAME attribs;
+
+collector_call_stmt -> target=APPLICATION_NAME "<-" target_endpoint=APPLICATION_NAME;
```
Comment (anzdaddy) [inline pkg/parser/endpoints.wbnf]: I don't think endpoints should support `::`. This is actually a good reason to split the bottom of the `APPLICATION_NAME` stack into a separate rule.

File: pkg/parser/sysl.wbnf [hunk 3]
```diff
@@ -85,7 +85,7 @@ PATH    -> /{[a-zA-Z0-9._-]+}:"/";
 DIGITS -> [1-9][0-9]*;
 TEXT -> [^(\r?\n)]+;
 TEXT_LINE -> /{\s* ( [^\s:(] [^(\r\n:]* [^\s\r\n(:] )};
-APPLICATION_NAME -> (pkg=@):"." > (app=@):"::" > (NAME | QSTRING)+;
+APPLICATION_NAME -> (pkg=@):"." > (app=@):"::" > (NAME | QSTRING | /{[a-zA-Z_][-\w]*(?:[-\t\_]*\w+)*})+;
```
Comment (anzdaddy) [inline pkg/parser/sysl.wbnf]:
```
APPNAME -> (pkg=@):"." > (app=@):"::" > APPNAME_PART;
APPNAME_PART -> (NAME | QSTRING | /{[a-zA-Z_][-\w]*(?:[-\t\_]*\w+)*})+;
```

Comment (anzdaddy) [inline pkg/parser/sysl.wbnf]: I'm guessing the regexp encompasses `NAME` and `STRING`, so those probably aren't required.

Comment (nofun97) [reply pkg/parser/sysl.wbnf]: the regexp is a more flexible `NAME`. `NAME` can only take alphanumeric while the regexp allows whitespace while still trimming the space.

Comment (nofun97) [reply pkg/parser/sysl.wbnf]: actually just realised `NAME` would always be used first. Will replace `NAME`.

File: pkg/parser/sysl.wbnf [hunk 4]
```diff
@@ -85,7 +85,8 @@ PATH    -> /{[a-zA-Z0-9._-]+}:"/";
 DIGITS -> [1-9][0-9]*;
 TEXT -> [^(\r?\n)]+;
 TEXT_LINE -> /{\s* ( [^\s:(] [^(\r\n:]* [^\s\r\n(:] )};
-APPLICATION_NAME -> (pkg=@):"." > (app=@):"::" > (NAME | QSTRING)+;
+APPLICATION_NAME -> (pkg=@):"." > (app=@):"::" > APPLICATION_NAME_PART;
+APPLICATION_NAME_PART -> (/{[a-zA-Z_][-\w]*(?:[-\t\_]*\w+)*} | QSTRING)+;
```
Comment (anzdaddy) [inline pkg/parser/sysl.wbnf]: I'd really much prefer `APPNAME`. It's pretty unambiguous and much more compact.

**Pull request merged**